[MNG-5965] Parallel build multiplies work if multiple goals are given

dbmeneses opened the pull request:
[mng-5965.zip](https://github.com/apache/maven/files/1124059/mng-5965.zip)
This p/r should also fix MNG-5705.

### Problem
When multiple tasks are given so that they are grouped in different task segments, the MultiThreadedBuilder fails. 
The reason is that it creates project segments, which is a set of unique pairs of (module, task).
It then handles each task separately, and it will schedules the build of the task for **all** root module in those pairs, instead of scheduling only the root modules which are associated with that particular task.

### Example
For example, given the project with the following hierarchy of modules:
```
root
-- module1
```

If we run `mvn task1 task2`, being task2 an aggregating task, we get the following project segments:
   (task1, root), (task1, module1), (task2, root).

The MultiThreadedBuilder will first handle task1, and will schedule the build of: (task1, root), (task1, root).
Each of these builds will then recursively build its children, meaning that _task1 will be executed twice for every module in the project_.

### Fix
This fix changes the builder so that it considers the project segments of each task segment separately, by creating a ConcurrencyDependencyGraph for each task segment.


### Related
I believe that MNG-5705 is a consequence of the same problem, because the builder will not count correctly the number of projects that are built because of the shared ConcurrencyDependencyGraph, resulting in a NPE.

### Test
I opened a separate P/R to add an IT for this issue: https://github.com/apache/maven-integration-testing/pull/22

It uses a very simple multi module project, which is attached here too.

To reproduce:
```
mvn clean install:help install -T10
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] module-1
[INFO] module-2
[INFO] base-project
[INFO] 
[INFO] Using the MultiThreadedBuilder implementation with a thread count of 10
[INFO] 
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building module-2 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building base-project 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ module-2 ---
[INFO] ------------------------------------------------------------------------
[INFO] Building module-1 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ base-project ---
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ module-1 ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building base-project 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ base-project ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building module-1 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ module-1 ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building module-2 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ module-2 ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building base-project 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ base-project ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] module-1 ........................................... SUCCESS [  0.069 s]
[INFO] module-2 ........................................... SUCCESS [  0.069 s]
[INFO] base-project ....................................... SUCCESS [  0.001 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.197 s (Wall Clock)
[INFO] Finished at: 2017-07-05T11:06:42+02:00
[INFO] Final Memory: 8M/240M
[INFO] ------------------------------------------------------------------------
[ERROR] NullPointerException
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
```

Note the clean goal being executed twice for each module, and the NPE (as described by MNG-5705).

Expected result, after the fix:
```
/tmp/apache-maven-3.5.1-SNAPSHOT/bin/mvn clean install:help install -T10
[INFO] Error stacktraces are turned on.
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] module-1
[INFO] module-2
[INFO] base-project
[INFO] 
[INFO] Using the MultiThreadedBuilder implementation with a thread count of 10
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building module-2 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ module-2 ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building base-project 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building module-1 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ module-1 ---
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ base-project ---
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Building base-project 0.1
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-install-plugin:2.4:help (default-cli) @ base-project ---
[INFO] Maven Install Plugin 2.4
  Copies the project artifacts to the user's local repository.

This plugin has 3 goals:

install:help
  Display help information on maven-install-plugin.
  Call mvn install:help -Ddetail=true -Dgoal=<goal-name> to display parameter
  details.

install:install
  Installs the project's main artifact, and any other artifacts attached by
  other plugins in the lifecycle, to the local repository.

install:install-file
  Installs a file in the local repository.
```

ifedorenko commented:
Can you provide a regression test that demonstrates the problem and the fix?

Semi-related, I recently fixed similar problem in [Takari Smart Builder](https://github.com/takari/takari-smart-builder) where aggregate tasks ran multiple times. Multithreading is hard :-)

dbmeneses commented:
@ifedorenko I've updated my previous comment. I added a simple reproducer and I submitted a P/R to maven-integration-testing with an IT.
BTW is it normal to have 100+ ITs failing with mvn 3.5.0?

ifedorenko commented:
No, having IT failures are not normal, and [jenkins seems to be happy](https://builds.apache.org/view/M-R/view/Maven%20Core%20ITs/job/core-integration-testing-maven-3-embedded/).

dbmeneses commented:
Ok my bad, I've managed to make them pass on my box as well.

ifedorenko commented:
The change looks reasonable but I don't use built-in multithreaded build support myself and do not feel comfortable merging this change without proper testing. I will have to defer to somebody more familiar with this part of Maven codebase.

dbmeneses commented:
@ifedorenko Ok, thanks for the help preparing the p/r. Is there anyone you could assign the review to?

ifedorenko commented:
No, sorry, I can't "assign" this to anyone, you just have to wait for somebody to review and merge this pull request.

FrzMe commented:
This PR is pretty old but it would be nice to get the referenced bugs fixed. 
@rfscholte @michael-o  it seems that you have worked with the relevant code before, could you review/merge?

michael-o commented:
I will have a look somewhere next week if no one is faster than me. @FrzMe please ping me next Wednesday or so if nothing happens. I will check the IT first for reproducibility.

FrzMe commented:
@michael-o I would be happy if you could have a look! Thank you so much

michael-o commented:
@dbmeneses why is there no parent element on the module POMs?

michael-o commented on maven-core/src/main/java/org/apache/maven/lifecycle/internal/builder/multithreaded/ConcurrencyDependencyGraph.java:
```diff
@@ -61,12 +61,12 @@ public int getNumberOfBuilds()
+    public Set<MavenProject> getRootSchedulableBuilds()
     {
-        List<MavenProject> result = new ArrayList<>();
+        Set<MavenProject> result = new HashSet<>();
```
I think this should be `LinkedHashSet` to retain Insertion order just like `ArrayList` does.

michael-o commented:
@dbmeneses the reason text contains some typos, for the sake of clarity and complexity of the issue, can you please fix them? It is hard enough to wrap one's mind around the actual problem. I had to read the problem several times to understand it. ;-)

jira-importer commented:
Resolve #7896
